### ISSUE TITLE
Strange error message when passing ill-typed code to Miri

### ISSUE THREAD

**Issue by RalfJung** [integer32llc/rust-playground#458]
[This example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b96edcfce8a9409c12d22c4b513229c3) shows the following error when hitting "Run":

```
   Compiling playground v0.0.1 (/playground)
error[E0425]: cannot find function `Droppable` in this scope
 --> src/main.rs:2:17
  |
2 |     let mut x = Droppable(0);
  |                 ^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `Droppable` in this scope
 --> src/main.rs:5:19
  |
5 |     unsafe { *p = Droppable(1); } // prints "dropping 0"
  |                   ^^^^^^^^^ not found in this scope

error: aborting due to 2 previous errors
```

But when I send it to Miri, I get
```
error[E0460]: found possibly newer version of crate `core` which `unicode_width` depends on
  |
  = note: perhaps that crate needs to be recompiled?
  = note: the following crate versions were found:
          crate `core`: /root/.cache/miri/HOST/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-00869b21179267d6.rlib
          crate `unicode_width`: /playground/target/debug/deps/libunicode_width-f959371afbea1e81.rlib

error: aborting due to previous error
```

**Comment by shepmaster** [comment]
> found possibly newer version of crate

This is for [any usage of Miri on the playground which includes a crate](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=da974c0b313093b5bce52c501170976c):

```
fn main() {
    memchr::memchr(0, &[]);
}
```

This has been present since Miri was introduced on the playground, AFAIR; TBH I assumed that Miri didn't handle crates yet.

**Comment by shepmaster** [comment]
I wonder if this shares commonalities with rust-lang/rust#57854, which is a similar "why is an external crate coming into play anyway".

**Comment by RalfJung** [comment]
But my example doesn't use a crate, does it?

Miri works fine with crates locally.

**Comment by RalfJung** [comment]
How do these other crates get built?  Maybe the issue is that they don't get built against Miri's sysroot.

**Comment by shepmaster** [comment]
> But my example doesn't use a crate, does it?

Right, and neither does [the linked issue](https://github.com/rust-lang/rust/issues/57854). Both just have the crates "hanging around".

**Comment by shepmaster** [comment]
> How do these other crates get built?

Miri [starts from the existing nightly container](https://github.com/integer32llc/rust-playground/blob/079b414a4620f9ef2028da57b7ac4f420afdcaff/compiler/miri/Dockerfile#L1). 

The base compiler containers [just run `cargo build`](https://github.com/integer32llc/rust-playground/blob/079b414a4620f9ef2028da57b7ac4f420afdcaff/compiler/base/Dockerfile#L44-L46)

> Maybe the issue is that they don't get built against Miri's sysroot.

Very likely. Does this mean that the Miri container needs to rebuild all the crates? Clippy [already does this](https://github.com/integer32llc/rust-playground/blob/079b414a4620f9ef2028da57b7ac4f420afdcaff/compiler/clippy/Dockerfile#L5-L10), so it might be time to reorganize the build process.

**Comment by RalfJung** [comment]
> Does this mean that the Miri container needs to rebuild all the crates?

Yes.  They have to be built against the sysroot in `~/.cache/miri/HOST`.

**Comment by shepmaster** [comment]
> They have to be built against the sysroot in `~/.cache/miri/HOST`.

What's the invocation for that? You're more familiar with our build process w.r.t. Miri.

**Comment by RalfJung** [comment]
You have a `Cargo.toml` will all the deps, right?  Doing `cargo clean && cargo miri run` in that project dir should be sufficient.

**Comment by shepmaster** [comment]
> `cargo clean`

Is this required? Docker has additive filesystems, deleting stuff from a previous layer doesn't save any space.

> `cargo miri run`

Presumably with at least a `fn main() {}`? There is no main.rs or lib.rs inside the container.

**Comment by shepmaster** [comment]
Implicitly fixed by #460 ?

**Comment by RalfJung** [comment]
Indeed, thanks :)
